tests/phpunit/tests/query/cacheResults.php

```diff
@@ -0,0 +1,707 @@
+			array(
+				'post_title'    => 'foo',
+				'post_password' => 'password',
+				'post_author'   => $user->ID,
+			)
+		);
+
+		$args   = array(
+			'cache_results' => true,
+			's'             => 'foo',
+		);
+		$query1 = new WP_Query();
+		$posts1 = $query1->query( $args );
+
+		wp_set_current_user( $user->ID );
+
+		$query2 = new WP_Query();
+		$posts2 = $query2->query( $args );
+		$this->assertEmpty( $posts1 );
+		$this->assertNotSame( $posts1, $posts2 );
+		$this->assertNotSame( $query1->found_posts, $query2->found_posts );
+	}
+
+	/**
+	 * @ticket 22176
+	 */
+	public function test_query_cache_new_comment() {
+		$args   = array(
+			'cache_results' => true,
+			'fields'        => 'ids',
+			'comment_count' => 1,
+		);
+		$query1 = new WP_Query();
+		$posts1 = $query1->query( $args );
+
+		self::factory()->comment->create( array( 'comment_post_ID' => self::$posts[0] ) );
+
+		$query2 = new WP_Query();
+		$posts2 = $query2->query( $args );
+
+		$this->assertNotSame( $posts1, $posts2 );
+		$this->assertContains( self::$posts[0], $posts2 );
+		$this->assertNotEmpty( $posts2 );
+		$this->assertNotSame( $query1->found_posts, $query2->found_posts );
+	}
+
+	/**
+	 * @ticket 22176
+	 */
+	public function test_main_comments_feed_includes_attachment_comments() {
+		$attachment_id = self::factory()->post->create( array( 'post_type' => 'attachment' ) );
+		$comment_id    = self::factory()->comment->create(
+			array(
+				'comment_post_ID'  => $attachment_id,
+				'comment_approved' => '1',
+			)
+		);
+
+		$args   = array(
+			'cache_results' => true,
+			'withcomments'  => 1,
+			'feed'          => 'feed',
+		);
+		$query1 = new WP_Query();
+		$query1->query( $args );
+
+		$query2 = new WP_Query();
+		$query2->query( $args );
+
+		$this->assertTrue( $query1->have_comments() );
+		$this->assertTrue( $query2->have_comments() );
+
+		$feed_comment = $query1->next_comment();
+		$this->assertEquals( $comment_id, $feed_comment->comment_ID );
+	}
+
+	/**
+	 * @ticket 22176
+	 */
+	public function test_query_cache_delete_comment() {
+		$comment_id = self::factory()->comment->create( array( 'comment_post_ID' => self::$posts[0] ) );
+		$args       = array(
+			'cache_results' => true,
+			'fields'        => 'ids',
+			'comment_count' => 1,
+		);
+		$query1     = new WP_Query();
+		$posts1     = $query1->query( $args );
+
+		wp_delete_comment( $comment_id, true );
+
+		$query2 = new WP_Query();
+		$posts2 = $query2->query( $args );
+
+		$this->assertNotSame( $posts1, $posts2 );
+		$this->assertEmpty( $posts2 );
+		$this->assertNotSame( $query1->found_posts, $query2->found_posts );
+	}
+
+	/**
+	 * @ticket 22176
+	 */
+	public function test_query_cache_update_post() {
+		$p1 = self::factory()->post->create();
+
+		$args   = array(
+			'cache_results' => true,
+			'fields'        => 'ids',
+		);
+		$query1 = new WP_Query();
+		$posts1 = $query1->query( $args );
+
+		wp_update_post(
+			array(
+				'ID'          => $p1,
+				'post_status' => 'draft',
+			)
+		);
```

Comment:
Once this moves to a shared fixture, would it be possible to update an existing comment -- just a few less non-SELECT DB queries.